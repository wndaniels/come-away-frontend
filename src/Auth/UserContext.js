import React from "react";
/**
 *  Context:
 *
 * Provides currentUser object and setter for it throughout app.
 */
const UserContext = React.createContext();

export default UserContext;
