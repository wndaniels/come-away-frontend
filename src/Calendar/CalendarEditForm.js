import React, { useContext, useEffect, useState } from "react";
import useLocalStorage from "../hooks/useLocalStorage.js";
import { useNavigate } from "react-router-dom";
import ComeAwayApi from "../Api/api.js";
import UserContext from "../Auth/UserContext.js";
import jwt from "jsonwebtoken";
import Alert from "../Common/Alert.js";
import DueDateEditForm from "../DueDate/DueDateEditForm.js";

export const TOKEN_STORAGE_ID = "comeaway-token";

const CalendarEditForm = () => {
  const navigate = useNavigate();
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  const [calAvailBegin, setCalAvailBegin] = useState([]);
  const [calAvailEnd, setCalAvailEnd] = useState([]);
  const [calByUser, setCalByUser] = useState([]);
  const [dueDateByUser, setDueDateByUser] = useState();
  const [formData, setFormData] = useState({
    businessBeginsHour: calByUser.businessBeginsHour,
    businessEndsHour: calByUser.businessEndsHour,
  });

  const [formError, setFormError] = useState([]);

  useEffect(
    function loadUserInfo() {
      async function getCurrentUser() {
        if (token) {
          try {
            let { username } = jwt.decode(token);
            ComeAwayApi.token = token;
            let currentUser = await ComeAwayApi.getCurrentUser(username);
            setCurrentUser(currentUser);
          } catch (errors) {
            setFormError(errors);
            setCurrentUser(null);
          }
        }
        setInfoLoaded(true);
      }
      setInfoLoaded(false);
      getCurrentUser();
    },
    [token, setCurrentUser]
  );

  useEffect(() => {
    async function getCalDataByUser() {
      const userCalData = await ComeAwayApi.getAllCals();
      try {
        userCalData.forEach((d) => {
          if (currentUser?.id === d.userId) setCalByUser(d);
        });
      } catch (errors) {
        return;
      }
    }
    getCalDataByUser([]);
  }, [currentUser?.id]);

  useEffect(() => {
    async function getDueDateByUser() {
      const dueDateData = await ComeAwayApi.getAllDueDates();

      try {
        dueDateData.forEach((d) => {
          if (currentUser?.id === d.userId) setDueDateByUser(d);
        });
      } catch (errors) {
        return;
      }
    }
    getDueDateByUser();
  }, [currentUser?.id]);

  useEffect(function getCalData() {
    async function getBeginHoursData() {
      try {
        const beginHourRes = await ComeAwayApi.getBeginHours();
        setCalAvailBegin(beginHourRes);
      } catch (errors) {
        return;
      }
    }

    async function getEndHourData() {
      try {
        const endHourRes = await ComeAwayApi.getEndHours();
        setCalAvailEnd(endHourRes);
      } catch (errors) {
        return;
      }
    }

    getBeginHoursData([]);
    getEndHourData([]);
  }, []);

  async function handleCalSubmit(evt) {
    evt.preventDefault();

    let calendarData = {
      businessBeginsHour: formData.businessBeginsHour,
      businessEndsHour: formData.businessEndsHour,
    };

    let username = currentUser.username;
    let id = calByUser.id;

    let updatedCal;

    try {
      updatedCal = await ComeAwayApi.updateCal(username, id, calendarData);
    } catch (errors) {
      setFormError([errors]);
      return;
    }
    navigate("/calendar");
    setFormData((f) => ({ ...f }));
    setFormError([]);
  }

  function handleIntChange(evt) {
    const { name, value } = evt.target;

    setFormData((f) => ({
      ...f,
      [name]: parseInt(value),
    }));

    setFormError([]);
  }

  async function handleDeleteCal(evt) {
    let username = currentUser.username;
    let id = calByUser.id;

    try {
      await ComeAwayApi.deleteCal(username, id);
    } catch (errors) {
      return;
    }
    navigate("/");
  }

  async function handleDeleteDueDate(evt) {
    let username = currentUser.username;
    let id = dueDateByUser.id;

    try {
      await ComeAwayApi.deleteDueDate(username, id);
    } catch (errors) {
      return;
    }
    navigate("/");
  }

  if (!infoLoaded) return;

  return (
    <div className="m-5">
      <div>
        <DueDateEditForm />
      </div>
      <div className="CalendarEditForm mt-5">
        <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
          <h1 className="mb-3">Edit Calendar</h1>
          <div className="card">
            <div className="card-body">
              <form method="patch">
                <div className="d-grid gap-3">
                  {formError.length ? (
                    <Alert
                      type="danger"
                      messages={["All fields must be complete."]}
                    ></Alert>
                  ) : null}

                  <div className="form-group">
                    <label htmlFor="businessBeginsHour">
                      * Select Start Time:
                    </label>
                    <select
                      name="businessBeginsHour"
                      className="form-control"
                      onChange={handleIntChange}
                    >
                      <option hidden>-</option>
                      {calAvailBegin &&
                        calAvailBegin.map((s, id) => (
                          <option value={s.businessBeginsHour} key={id}>
                            {s.hourTitle}
                          </option>
                        ))}
                    </select>
                  </div>

                  <div className="form-group">
                    <label htmlFor="businessEndsHour">* Select End Time:</label>
                    <select
                      name="businessEndsHour"
                      className="form-control mb-3"
                      onChange={handleIntChange}
                    >
                      <option hidden>-</option>
                      {calAvailEnd &&
                        calAvailEnd.map((e, id) => (
                          <option
                            defaultValue={e}
                            value={e.businessEndsHour}
                            key={id}
                          >
                            {e.hourTitle}
                          </option>
                        ))}
                    </select>
                  </div>
                </div>

                <button
                  onClick={handleCalSubmit}
                  className="btn btn-sm btn-primary "
                >
                  Update Calendar
                </button>
                <button
                  onClick={() => {
                    handleDeleteCal();
                    handleDeleteDueDate();
                  }}
                  className="btn btn-sm btn-danger ms-3"
                >
                  Delete
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default CalendarEditForm;
