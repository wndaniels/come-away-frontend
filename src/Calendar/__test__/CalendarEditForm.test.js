import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom";
import CalendarEditForm from "../CalendarEditForm";
import UserContext from "../../Auth/UserContext";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => jest.fn(),
}));

describe("CalendarEditForm", () => {
  it("renders without crashing", () => {
    const { asFragment } = render(
      <Wrapper>
        <CalendarEditForm />
      </Wrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("can find placeholder", () => {
    const { getByRole, getByText } = render(
      <Wrapper>
        <CalendarEditForm />
      </Wrapper>
    );

    const heading = getByRole("heading", {
      name: /edit calendar/i,
      class: "mb-3",
    });
    const deleteButtonText = getByText(/delete/i);

    expect(heading).toBeInTheDocument();
    expect(deleteButtonText).toBeInTheDocument();
  });
});

function Wrapper({ children }) {
  return (
    <UserContext.Provider
      value={{
        currentUser: jest.fn(),
        setCurrentUser: jest.fn(),
        login: jest.fn(),
        signup: jest.fn(),
        logout: jest.fn(),
      }}
    >
      {children}
    </UserContext.Provider>
  );
}
