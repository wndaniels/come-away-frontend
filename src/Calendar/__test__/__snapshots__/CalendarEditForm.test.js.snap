// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarEditForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="m-5"
  >
    <div>
      <div
        class="DueDateEditForm"
      >
        <div
          class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
        >
          <h1>
            Edit Due Date
          </h1>
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <form
                method="post"
              >
                <div
                  class="d-grid gap-3"
                >
                  <div
                    class="form-group"
                  >
                    <label>
                      Baby's Name:
                    </label>
                    <input
                      class="form-control"
                      name="babyName"
                      type="text"
                      value=""
                    />
                  </div>
                  <div
                    class="form-group"
                  >
                    <label>
                      Month
                    </label>
                    <select
                      class="form-control"
                      name="monthId"
                    >
                      <option
                        hidden=""
                      >
                        -
                      </option>
                    </select>
                  </div>
                  <div
                    class="form-group"
                  >
                    <label>
                      Day
                    </label>
                    <select
                      class="form-control"
                      name="dayId"
                    >
                      <option
                        hidden=""
                      >
                        -
                      </option>
                    </select>
                  </div>
                  <div
                    class="form-group"
                  >
                    <label>
                      Year
                    </label>
                    <select
                      class="form-control"
                      name="yearId"
                    >
                      <option
                        hidden=""
                      >
                        -
                      </option>
                    </select>
                  </div>
                </div>
                <button
                  class="btn btn-sm btn-primary mt-3"
                >
                  Update Due Date
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="CalendarEditForm mt-5"
    >
      <div
        class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
      >
        <h1
          class="mb-3"
        >
          Edit Calendar
        </h1>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <form
              method="patch"
            >
              <div
                class="d-grid gap-3"
              >
                <div
                  class="form-group"
                >
                  <label
                    for="businessBeginsHour"
                  >
                    * Select Start Time:
                  </label>
                  <select
                    class="form-control"
                    name="businessBeginsHour"
                  >
                    <option
                      hidden=""
                    >
                      -
                    </option>
                  </select>
                </div>
                <div
                  class="form-group"
                >
                  <label
                    for="businessEndsHour"
                  >
                    * Select End Time:
                  </label>
                  <select
                    class="form-control mb-3"
                    name="businessEndsHour"
                  >
                    <option
                      hidden=""
                    >
                      -
                    </option>
                  </select>
                </div>
              </div>
              <button
                class="btn btn-sm btn-primary "
              >
                Update Calendar
              </button>
              <button
                class="btn btn-sm btn-danger ms-3"
              >
                Delete
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
